activerecord/test/cases/transactions_test.rb

```diff
@@ -679,6 +679,34 @@ def test_restore_frozen_state_after_double_destroy
     assert_not_predicate topic, :frozen?
   end
 
+  def test_restore_new_record_after_double_save
+    topic = Topic.new
+
+    Topic.transaction do
+      topic.save!
+      topic.save!
+      raise ActiveRecord::Rollback
+    end
+
+    assert_predicate topic, :new_record?
+  end
+
+  def test_dont_restore_new_record_in_subsequent_transaction
+    topic = Topic.new
+
+    Topic.transaction do
+      topic.save!
+      topic.save!
+    end
+
+    Topic.transaction do
+      topic.save!
+      raise ActiveRecord::Rollback
+    end
+
+    assert_predicate topic, :persisted?
```

Comment:
What does this test? This test case will pass without this PR.

Reply:
If the parts of this PR related to rollbacks are applied on their own, this test will start to fail; in other words, it demonstrates that [this new line](https://github.com/rails/rails/commit/73e8519915f82ff44956f04d22841e2ec8a48b38#diff-174733f2db65ef1bc53e3222d6ac0e61R476) is necessary.

Alternatively, this test will fail on master if the final assertion is duplicated.

Reply:
👍 done!